augur/metadata_file.py

```diff
@@ -0,0 +1,89 @@
+import functools
+from os import access, R_OK
+import pandas
+
+
+class MetadataFile:
+    """
+    Represents a CSV or TSV file containing metadata
+
+    The file must contain exactly one of a column named `strain` or `name`,
+    which is used to match metadata with samples.
+    """
+
+    def __init__(self, fname, query):
+        self.fname = fname
+        self.query = query
+
+        self.key_type = self.find_key_type()
+
+        self.check_file()
+
+    def check_file(self):
+        if not access(self.fname, R_OK):
+            raise ValueError(
+                f"Metadata file {self.fname} does not exist or is not readable"
+            )
+
+    def read(self):
+        meta_dict = {}
+        for row in self.metadata:
+            key_val = row[self.key_type]
+
+            if key_val in meta_dict:
+                raise ValueError(
+                    f"Duplicate {self.key_type} in metadata files: `{key_val}`"
+                )
+
+            meta_dict[key_val] = row
+
+        return meta_dict, self.columns
+
+    @property
+    def metadata(self):
+        """
+        Return list of dicts representing the metadata in the file.
+
+        If a query was supplied, apply it.
+        """
+
+        metadata = self.parse_file()
+
+        if self.query:
+            try:
+                metadata = metadata.query(self.query)
+            except ValueError:
+                raise ValueError(
+                    f"Error applying pandas query to metadata: `{self.query}`"
+                )
+
+        return [row[1].to_dict() for row in metadata.iterrows()]
```

Comment:
pandas's `iterrows` method is very slow for some reason that I've never fully understood. Reading a ~70,000 line metadata file (for our seasonal flu data) takes an average of 10 +/- 2 seconds with this loop.

I think we can collapse the logic from the `metadata` property and the `read` method into a single method that will also run faster. Here is a general approach I had worked out a while back and never got around to committing:

```python
# Define strain key (ideally from data).
strain_key = "strain"

# Save columns.
metadata_columns = list(metadata.columns)

# Find duplicates.
duplicates = metadata[strain_key].value_counts().reset_index().query("strain > 1")["index"].values

if len(duplicates) > 0:
    raise ValueError(f"Duplicate values from column {strain_key} found: {', '.join(duplicates)}")

# Index data frame by the strain key.
metadata["_index"] = metadata[strain_key]

# Convert the data frame to a dictionary of dictionaries 
# indexed by strain key.
metadata_dict = metadata.set_index("_index").to_dict("index")

return metadata_dict, metadata_columms
```

The trick of creating a new "_index" column, indexing by this, and running `to_dict on the entire data frame speeds up the conversion of rows to dictionaries substantially. For the same seasonal flu data mentioned above, this `metadata_dict = ...` line runs in 575 +/- 9 milliseconds.

The approach to looking for duplicates with `value_counts` also allows us to use pandas's optimized logic for this type of operation with a single line of code.

Reply:
Just a wild guess: `iterrows` might be slow because Pandas uses column-oriented data structures, not row-oriented ones. For a row-wise operation like `iterrows`, it has to pull each column's value from a different non-adjacent data structure.

Reply:
How important is speed here? If it's a priority, we're likely to get the best performance (a completely untested guess) by ditching Pandas, using Python's `csv` module, and implementing the query logic ourselves. It might even be more readable, since it would not require understanding of Pandas (compared to the `_index` trick). The cost, of course, is additional code in the class.

The built-in CSV module even provides `csv.Sniffer` for delimiter autodetection, which may be useful in #574.

Reply:
It's probably also worth considering the `engine="numexpr"` argument to `DataFrame.query`. I don't have data handy to benchmark it compared to the default `python` engine, but https://pandas.pydata.org/pandas-docs/stable/user_guide/enhancingperf.html#pandas-eval-performance suggests it may help.

Reply:
These are good points, @elebow! Speed is less important than readability, for sure. In this case the "trick" is really more of a pandas idiom. In the same way you'd work with data in R using vectors instead of for loops, the `to_dict` example above takes advantage of pandas's column-based approach (as @tsibley correctly guessed ;) ). We'd like to stick with pandas here over the built-in `csv` module to take advantage of the other features pandas offers like the `query` interface, type inference for columns, data grouping support, etc..

We can write the pandas code above with clearer names like this:

```python
# Index data frame by the strain key.
metadata["strain_key"] = metadata[strain_key]

# Convert the data frame to a dictionary of dictionaries indexed by strain key.
metadata_dict = metadata.set_index("strain_key").to_dict("index")
```

I know this a slightly subjective preference here, but I find this kind of pandas code just as readable as the current list comprehension with the additional benefit of being an order of magnitude faster.

The duplicates finding code in the example above also has the benefit of reporting all duplicates in one message, avoiding the situation where the user has to iterate through multiple manual duplicate removals. This approach could be helpful in the near future if we choose to [handle duplicate sequences a little more gracefully than we do now](https://github.com/nextstrain/augur/issues/586).

Since the `metadata` property is only referenced from the `read` method, I'm still in favor of merging these into the `read` method, for now. We can always refactor these out in the future, if we have the need. I do really appreciate your focus on implementing this functionality in proper Python classes, though. It's something Augur might benefit from more of in the future.

Reply:
Makes sense. I added a comment explaining the duplicated column, but maybe it's not necessary if this is a common idiom.